debian/pyversions.1
.TH PYVERSIONS
.SH NAME
pyversions \- print python version information
.SH SYNOPSIS
.PP
.B pyversions
.I [-h] [-v] [<options>]
.SH DESCRIPTION
.PP
.B pyversions
prints information about installed, supported python runtimes, the
default runtime, and parses the information of the PythonVersion
fields in the package control file.
.SH OPTIONS
.TP
.I -d, --default
Show the default python version.
.TP
.I -s, --supported
Show the supported python versions.
.TP
.I -r, --requested <version string>|<control file>
Reads the value of the
.B XS-Python-Version
field in the source section of a control file and shows all matching
python versions. The parameter is interpreted as a version string, if
it is not the name of a file. If the XS-Python-Version field is
missing, get the version information from debian/pyversions.
Without any parameter, it will fall back to list the supported Python
versions after checking debian/control and debian/pyversions.
.TP
.I -i, --installed
Show the installed supported python versions.
.TP
.I -v, --version
Limit the output to the version numbers of the python versions.
.TP
.I -h, --help
Print a help text.
.SH SEE ALSO
Python policy.
.SH AUTHOR
Matthias Klose


debian/changelog
python-defaults (2.6.5-2) UNRELEASED; urgency=low

  * 

 -- Matthias Klose <doko@debian.org>  Tue, 18 May 2010 12:28:25 +0200

python-defaults (2.6.5-1) experimental; urgency=low

  [ Matthias Klose ]
  * Bump version to 2.6.5.
  * Make python2.6 the default Python version.
  * Build a python-doc package.
  * python: Update the valgrind file to 2.6.

  [ Piotr Ozarowski ]
  * python-minimal: Install pycompile and pyclean binaries.
  * Add lsb-release to Build-Depends
  * Add dh_python2 stuff

 -- Matthias Klose <doko@debian.org>  Fri, 14 May 2010 18:50:21 +0200

python-defaults (2.5.4-9) unstable; urgency=low

  * python-all-dev: Fix typo in dependency list. Closes: #566278.

 -- Matthias Klose <doko@debian.org>  Fri, 22 Jan 2010 18:38:31 +0100

python-defaults (2.5.4-6) unstable; urgency=low

  * Add python2.6 as a supported Python version.
  * python-all{,-dev,-dbg}: Depend on the 2.6 packages.
  * Run the rtinstall script, when 2.5 became a supported version (2.4.4-3)
    and python2.5 is installed.
  * Run the rtinstall script, when 2.6 became a supported version (2.5.4-6)
    and python2.6 is installed. Closes: #565952.
  * Fix reference to list of modules in the python-minimal package.
    Closes: #558403.

 -- Matthias Klose <doko@debian.org>  Sun, 17 Jan 2010 01:38:21 +0100

python-defaults (2.5.4-5) unstable; urgency=low

  * Fix typos from policy update (Closes: #561091, #561093)
  * Drop python2.4 as a supported Python
    - Remove python2.4 depends and update descriptions for python-all,
      python-all-dev, and python-all-dbg
    - Drop python2.4 from supported versions and add to old and unsupported
      versions
    - Update pyversions.py to drop python2.4 from the supported list
  * Correct path to Konsole Python desktop file (Closes: #381659)

 -- Scott Kitterman <scott@kitterman.com>  Wed, 23 Dec 2009 19:59:15 -0500

python-defaults (2.5.4-4) unstable; urgency=low

  [ Scott Kitterman ]
  * Update Python Policy to match current packaging practice:
    - Major overall update.
    - Remove extraneous paragraph from Provides section. (Closes: #415213)
    - Remove obsolete mentions of dh_python. (Closes: #527009)
    - Fix policy typo. (Closes: #523226)
    - Add pointer to doc of the site module for site-specific paths.
      (Closes: #287197)
    - Correct package dependency rules (Closes: #379455)
  * Move debhelper from build-depends-indep to build-depends to satisfy clean
    requirements.
  * Drop unneeded build-depends-indep on python.
  * Bump standards version to 3.8.3.
  * Drop deprecated dh_desktop from debian/rules.
  * Move dbg packages to section debug.
  * Make python-all-dbg priority extra.
  * Remove redundant debian/FAQ.html.
  * Update (= ${Source-Version}) depends to (>= ${binary:Version}).
  * Correct Idle depends and description (Closes: #482163).
  * Update python version reference in README.Debian (Closes: #494068).
  * Update Debian menu section for idle.
  * Fix capitalization of Python in the python-dev descriptions.
  * Add myself to uploaders.

  [ Matthias Klose ]
  * python-{dev,dbg}-dev: Add bogus dependencies on python to work around
    lintian's broken usr-share-doc-symlink-without-dependency check.

 -- Scott Kitterman <scott@kitterman.com>  Mon, 14 Dec 2009 06:03:32 -0500

python-defaults (2.5.4-2) unstable; urgency=high

  * python-all*: Tighten the dependencies on the python-* packages.

 -- Matthias Klose <doko@debian.org>  Thu, 26 Feb 2009 12:44:47 +0000

python-defaults (2.5.4-1) unstable; urgency=medium

  * Depend on python2.5-2.5.4.
  * python: Install /usr/share/python/python.mk, helper functions for
    getting the python library dir and install args for distutils installs.
  * Accept initial blank line in debian/control when parsing version
    information.

 -- Matthias Klose <doko@debian.org>  Wed, 25 Feb 2009 02:19:09 +0000

python-defaults (2.5.2-3) unstable; urgency=low

  * python: Provide python-ctypes and python-wsgiref.

 -- Matthias Klose <doko@debian.org>  Sat, 01 Nov 2008 09:15:19 +0100

python-defaults (2.5.2-2) unstable; urgency=low

  * python: Provide python-plistlib.
  * python-minimal: Recommend python (suggested by Neil McGovern).

 -- Matthias Klose <doko@debian.org>  Sat, 26 Jul 2008 02:37:00 +0200

python-defaults (2.5.2-1) unstable; urgency=low

  * Maintainer upload; asking the Release Team to better coordinate
    things. Being in contact with two members of the release team for
    two or three weeks when to upload without getting in the way of
    other transitions, an unannounced upload of a third member of this
    team is at least surprising.
  * debian/python.postinst.in: Fix the bug introduced in the NMU, run the
    rtupdate hooks when upgrading from a version < 2.5.2-1. Closes: #476095.

 -- Matthias Klose <doko@debian.org>  Thu, 17 Apr 2008 20:32:59 +0200

python-defaults (2.5.2-0.1) unstable; urgency=low

  * NMU by the Release Team not to lose the sweet spot for the Python
    transition.
  * Make python2.5 the default Python version.

 -- Adeodato Simó <dato@net.com.org.es>  Sun, 13 Apr 2008 11:29:32 +0200

python-defaults (2.4.4-6) unstable; urgency=low

  * python.postinst: Fix cleanup of temporary files. Closes: #424703, #424704.

 -- Matthias Klose <doko@debian.org>  Thu, 17 May 2007 08:17:16 +0200

python-defaults (2.4.4-5) unstable; urgency=low

  * Run the python2.5 rtinstall hooks, if python2.5 was already
    installed as an unsupported version.

 -- Matthias Klose <doko@debian.org>  Wed, 16 May 2007 21:19:51 +0200

python-defaults (2.4.4-4) unstable; urgency=low

  * New package python-all-dbg.

 -- Matthias Klose <doko@debian.org>  Fri, 04 May 2007 06:56:27 +0200

python-defaults (2.4.4-3) unstable; urgency=low

  * Add python2.5 as a supported python version.
  * Depend on python2.4 packages (>= 2.4.4-4), built for ldbl128 on
    alpha, powerpc, s390, sparc.
  * python-all*: Depend on python2.5 packages (>= 2.4.5-1), built
    for ldbl128 on alpha, powerpc, s390, sparc.

debian/pyversions.py
#! /usr/bin/python

import os, re, sys
try:
    SetType = set
except NameError:
    import sets
    SetType = sets.Set
    set = sets.Set

_defaults = None
def read_default(name=None):
    global _defaults
    from ConfigParser import SafeConfigParser, NoOptionError
    if not _defaults:
        if os.path.exists('/usr/share/python/debian_defaults'):
            config = SafeConfigParser()
            config.readfp(file('/usr/share/python/debian_defaults'))
            _defaults = config
    if _defaults and name:
        try:
            value = _defaults.get('DEFAULT', name)
        except NoOptionError:
            raise ValueError
        return value
    return None

def parse_versions(vstring, add_exact=False):
    import operator
    operators = { None: operator.eq, '=': operator.eq,
                  '>=': operator.ge, '<=': operator.le,
                  '<<': operator.lt
                  }
    vinfo = {}
    exact_versions = set([])
    version_range = set(supported_versions(version_only=True)
                        + old_versions(version_only=True))
    relop_seen = False
    for field in vstring.split(','):
        field = field.strip()
        if field == 'all':
            vinfo['all'] = 'all'
            continue
        if field in ('current', 'current_ext'):
            vinfo['current'] = field
            continue
        vinfo.setdefault('versions', set())
        ve = re.compile('(>=|<=|<<|=)? *(\d\.\d)$')
        m = ve.match(field)
        try:
            op, v = m.group(1), m.group(2)
            if op in (None, '='):
                exact_versions.add(v)
            else:
                relop_seen = True
                filtop = operators[op]
                version_range = [av for av in version_range if filtop(av ,v)]
        except Exception:
            raise ValueError, 'error parsing Python-Version attribute'
    if add_exact:
        if exact_versions:
            vinfo['vexact'] = exact_versions
        if 'versions' in vinfo:
            if relop_seen:
                vinfo['versions'] = set(version_range)
            else:
                del vinfo['versions']
    else:
        if 'versions' in vinfo:
            vinfo['versions'] = exact_versions
            if relop_seen:
                vinfo['versions'] = exact_versions.union(version_range)
    return vinfo

_old_versions = None
def old_versions(version_only=False):
    global _old_versions
    if not _old_versions:
        try:
            value = read_default('old-versions')
            _old_versions = [s.strip() for s in value.split(',')]
        except ValueError:
            _old_versions = []
    if version_only:
        return [v[6:] for v in _old_versions]
    else:
        return _old_versions

_unsupported_versions = None
def unsupported_versions(version_only=False):
    global _unsupported_versions
    if not _unsupported_versions:
        try:
            value = read_default('unsupported-versions')
            _unsupported_versions = [s.strip() for s in value.split(',')]
        except ValueError:
            _unsupported_versions = []
    if version_only:
        return [v[6:] for v in _unsupported_versions]
    else:
        return _unsupported_versions

_supported_versions = None
def supported_versions(version_only=False):
    global _supported_versions
    if not _supported_versions:
        try:
            value = read_default('supported-versions')
            _supported_versions = [s.strip() for s in value.split(',')]
        except ValueError:
            cmd = ['/usr/bin/apt-cache', '--no-all-versions',
                   'show', 'python-all']
            try:
                import subprocess
                p = subprocess.Popen(cmd, bufsize=1,
                                     shell=False, stdout=subprocess.PIPE)
                fd = p.stdout
            except ImportError:
                fd = os.popen(' '.join(cmd))
            depends = None
            for line in fd:
                if line.startswith('Depends:'):
                    depends = line.split(':', 1)[1].strip().split(',')
            fd.close()
            if depends:
                depends = [re.sub(r'\s*(\S+)[ (]?.*', r'\1', s) for s in depends]
                _supported_versions = depends
            if not _supported_versions:
                # last resort: python-minimal not installed, apt-cache
                # not available, hard code the value, #394084
                _supported_versions = ['python2.5',]
    if version_only:
        return [v[6:] for v in _supported_versions]
    else:
        return _supported_versions

_default_version = None
def default_version(version_only=False):
    global _default_version
    if not _default_version:
        try:
            _default_version = link = os.readlink('/usr/bin/python')
        except OSError:
            _default_version = None
            try:
                cmd = ['/usr/bin/python', '-c', 'import sys; print sys.version[:3]']
                import subprocess
                p = subprocess.Popen(cmd, bufsize=1,
                                     shell=False, stdout=subprocess.PIPE)
                fd = p.stdout
            except ImportError:
                fd = os.popen("/usr/bin/python -c 'import sys; print sys.version[:3]'")
            line = fd.readline().strip()
            fd.close()
            if re.match(r'\d\.\d$', line):
                _default_version = 'python' + line
        # consistency check
        try:
            debian_default = read_default('default-version')
        except ValueError:
            debian_default = "python2.5"
        if not _default_version in (debian_default, os.path.join('/usr/bin', debian_default)):
            raise ValueError, "/usr/bin/python does not match the python default version. It must be reset to point to %s" % debian_default
        _default_version = debian_default
    if version_only:
        return _default_version[6:]
    else:
        return _default_version

def requested_versions(vstring, version_only=False):
    versions = None
    vinfo = parse_versions(vstring, add_exact=True)
    supported = supported_versions(version_only=True)
    if len(vinfo) == 1:
        if 'all' in vinfo:
            versions = supported
        elif 'current' in vinfo:
            versions = [default_version(version_only=True)]
        elif 'vexact' in vinfo:
            versions = vinfo['vexact']
        else:
            versions = vinfo['versions'].intersection(supported)
    elif 'all' in vinfo and 'current' in vinfo:
        raise ValueError, "both `current' and `all' in version string"
    elif 'all' in vinfo:
        if 'versions' in vinfo:
            versions = vinfo['versions'].intersection(supported)
        else:
            versions = set(supported)
        if 'vexact' in vinfo:
            versions.update(vinfo['vexact'])
    elif 'current' in vinfo:
        current = default_version(version_only=True)
        if not current in vinfo['versions']:
            raise ValueError, "`current' version not in supported versions"
        versions = [current]
    elif 'versions' in vinfo or 'vexact' in vinfo:
        versions = set()
        if 'versions' in vinfo:
            versions = vinfo['versions'].intersection(supported)
        if 'vexact' in vinfo:
            versions.update(vinfo['vexact'])
    else:
        raise ValueError, 'error in version string'
    if not versions:
        raise PyCentralEmptyValueError, 'empty set of versions'
    if version_only:
        return versions
    else:
        return ['python%s' % v for v in versions]

def installed_versions(version_only=False):
    import glob
    supported = supported_versions()
    versions = [os.path.basename(s)
                for s in glob.glob('/usr/bin/python[0-9].[0-9]')
                if os.path.basename(s) in supported]
    versions.sort()
    if version_only:
        return [v[6:] for v in versions]
    else:
        return versions

class ControlFileValueError(ValueError):
    pass
class MissingVersionValueError(ValueError):
    pass

def extract_pyversion_attribute(fn, pkg):
    """read the debian/control file, extract the XS-Python-Version
    field; check that XB-Python-Version exists for the package."""

    version = None
    sversion = None
    section = None
    for line in file(fn):
        line = line.strip()
        if line == '':
            if section == None:
                continue
            if pkg == 'Source':
                break
            section = None
        elif line.startswith('Source:'):
            section = 'Source'
        elif line.startswith('Package: ' + pkg):
            section = pkg
        elif line.startswith('XS-Python-Version:'):
            if section != 'Source':
                raise ValueError, \
                      'attribute XS-Python-Version not in Source section'
            sversion = line.split(':', 1)[1].strip()
        elif line.startswith('XB-Python-Version:'):
            if section == pkg:
                version = line.split(':', 1)[1].strip()
    if section == None:
        raise ControlFileValueError, 'not a control file'
    if pkg == 'Source':
        if sversion == None:
            raise MissingVersionValueError, \
                  'missing XS-Python-Version in control file'
        return sversion
    if version == None:
        raise MissingVersionValueError, \
              'missing XB-Python-Version for package `%s' % pkg
    return version

# compatibility functions to parse debian/pyversions

def version_cmp(ver1,ver2):
    v1=[int(i) for i in ver1.split('.')]
    v2=[int(i) for i in ver2.split('.')]
    return cmp(v1,v2)

def requested_versions_bis(vstring, version_only=False):
    versions = []
    py_supported_short = supported_versions(version_only=True)
    for item in vstring.split(','):
        v=item.split('-')
        if len(v)>1:
            if not v[0]:
                v[0] = py_supported_short[0]
            if not v[1]:
                v[1] = py_supported_short[-1]
            for ver in py_supported_short:
                try:
                    if version_cmp(ver,v[0]) >= 0 \
                           and version_cmp(ver,v[1]) <= 0:
                        versions.append(ver)
                except ValueError:
                    pass
        else:
            if v[0] in py_supported_short:
                versions.append(v[0])
    versions.sort(version_cmp)
    if not versions:
        raise ValueError, 'empty set of versions'
    if not version_only:
        versions=['python'+i for i in versions]
    return versions

def extract_pyversion_attribute_bis(fn):
    vstring = file(fn).readline().rstrip('\n')
    return vstring

def main():
    from optparse import OptionParser
    usage = '[-v] [-h] [-d|--default] [-s|--supported] [-i|--installed] [-r|--requested <version string>|<control file>]'
    parser = OptionParser(usage=usage)
    parser.add_option('-d', '--default',
                      help='print the default python version',
                      action='store_true', dest='default')
    parser.add_option('-s', '--supported',
                      help='print the supported python versions',
                      action='store_true', dest='supported')
    parser.add_option('-r', '--requested',
                      help='print the python versions requested by a build; the argument is either the name of a control file or the value of the XS-Python-Version attribute',
                      action='store_true', dest='requested')
    parser.add_option('-i', '--installed',
                      help='print the installed supported python versions',
                      action='store_true', dest='installed')
    parser.add_option('-v', '--version',
                      help='print just the version number(s)',
                      default=False, action='store_true', dest='version_only')
    opts, args = parser.parse_args()
    program = os.path.basename(sys.argv[0])

    if opts.default and len(args) == 0:
        try:
            print default_version(opts.version_only)
        except ValueError, msg:
            print "%s:" % program, msg
            sys.exit(1)
    elif opts.supported and len(args) == 0:
        print ' '.join(supported_versions(opts.version_only))
    elif opts.installed and len(args) == 0:
        print ' '.join(installed_versions(opts.version_only))
    elif opts.requested and len(args) <= 1:
        if len(args) == 0:
            versions = 'debian/control'
        else:
            versions = args[0]
        try:
            if os.path.isfile(versions):
                fn = versions
                try:
                    vstring = extract_pyversion_attribute(fn, 'Source')
                    vs = requested_versions(vstring, opts.version_only)
                except ControlFileValueError:
                    sys.stderr.write("%s: not a control file: %s, " \
                                     % (program, fn))
                    sys.exit(1)
                except MissingVersionValueError:
                    fn = os.path.join(os.path.dirname(fn), 'pyversions')
                    sys.stderr.write("%s: missing XS-Python-Version in control file, fall back to %s\n" \
                                     % (program, fn))
                    try:
                        vstring = extract_pyversion_attribute_bis(fn)
                        vs = requested_versions_bis(vstring, opts.version_only)
                    except IOError:
                        sys.stderr.write("%s: missing debian/pyversions file, fall back to supported versions\n" \
                                         % program)
                        vs = supported_versions(opts.version_only)
            else:
                vs = requested_versions(versions, opts.version_only)
            print ' '.join(vs)
        except ValueError, msg:
            sys.stderr.write("%s: %s\n" % (program, msg))
            sys.exit(1)
    else:
        sys.stderr.write("usage: %s %s\n" % (program, usage))
        sys.exit(1)

if __name__ == '__main__':
    main()
